### ISSUE TITLE
Design how to integrate Hummingbot with external signal providers

### ISSUE THREAD

**Issue by mifeng** [CoinAlpha/hummingbot#2405]
(Originally Clubhouse story [9513](https://app.clubhouse.io/coinalpha/story/9513/design-how-to-integrate-hummingbot-with-external-signal-providers))

# Why
We are seeing increasing interest in integrating external signals or controls into Hummingbot. Some use cases that users want to enable are:
- integrating TradingView (most popular trader tool) with Hummingbot: https://coincise.co/how-to-send-commands-from-tradingview-to-hummingbot-telegram-account/
- integrating signals from analytics providers like IntoTheBlock into Hummingbot
- integrating [Twitch extensions](https://www.twitch.tv/p/extensions/) into Hummingbot to enable more streaming (our user QcPat has already started a [channel to steam Hummingbot](https://www.twitch.tv/crypto_trade)

The Twitch use case is less important than the other two, since we're not sure if streaming is something we want to promote.

# What
Design an approach to allow developers to integrate external signal providers like this into Hummingbot
- this is aimed at developers so no UI is needed
- signals can either be automated (predictive signals) or manual (Twitch/TradingView)
- since this will be an important long term aspect of the Hummingbot codebase, I think we should take the time to take through the architecture carefully

**Comment by mifeng** [comment]
@bitphage Clubhouse comment:

Some thoughts on API design.

HTTP REST connector ---> BaseAPI
Telegram connector ----> BaseAPI ----> End Handler

An API connector is responsible for exposing API via some protocol - http, telegram, etc. It then proxies all requests into BaseAPI which in turn determines a correct next handler. All API calls could be divided into several groups and routed accordingly to the proper handlers. The base API groups could be:

* Core API: start/stop/create/delete a bot, or modify global config, get global stats etc
* Strategy API: execute anything on strategy level or control strategy behavior, for example user may do a call "execute stop loss now" etc
* Script API: provide any custom data into user-supplied script (it supposed to have handlers), this is where literally any integration is possible.

Each API connector could be executed as a separate asyncio Task:

* For http, a [sanic](https://sanic.readthedocs.io/en/latest/) could be used.
* For telegram, [aiogram](https://github.com/aiogram/aiogram)

DEXBot plugin: auto-discover everything from dexbot.plugins namespace and run it, every plugin as separate Task. A plugin could do anything like check for new version, report stats, track portfolio valuation or order a pizza. Here is how example plugin could look: https://github.com/bitphage/DEXBot/blob/500-plugin-infrastructure/dexbot/plugins/dummy.py.example

**Comment by phbrgnomo** [comment]
An alternative to having the bot read signals from Tradingview is to create a workframe that allow users to create their own indicators using the data already being collected from the exchanges.

This could work as a module of scripts, but as a separated entity.

The data flow would be something like:

The main bot module collect the data -> TA module use the data to calculate the Indicator (like scripts do) -> Script use the return data from TA module to send the modifications to the bot.

For professionals, this could be a more interesting and efficient way to apply Indicators, since the data is collected directly from the source (the exchange) and also allow the trader to implement their indicators based on tick data, instead of candles, and reduce the processing time.

Depending on the pair and the exchange, i have seen some discrepancies from Tradingview and the data of an exchange (like volume).

Another advantage would be that advanced users could create their Indicators using their own customized information source, instead of only price and volume, allowing the implementation of more quant strategies.

Also, since this would be on python, a separate module for Technical Indicators also allows the use of the vast library of code already available on the web.

Two things would be needed for this architecture:

- The bot would need to have a way to store and call more information from the exchange to feed the TA module (volume per tick, for example)
- The bot would need to transform the tick data into common format used by indicators (transform tick data do candles (OHLC table))
- Similar to scripts, we would need a documentation hub to show the functions as a reference guide for developers.

**Comment by bitphage** [comment]
@phbrgnomo yeah having own indicators is cool, I would research how it's done in freqtrade bot, they have TA-lib integrated, see https://www.freqtrade.io/en/latest/strategy-customization/

**Comment by TheHolyRoger** [comment]
Not sure how much work you guys have already done on this but I'd suggest Quart over Sanic

**Comment by krisj** [comment]
> Not sure how much work you guys have already done on this but I'd suggest Quart over Sanic

I don't have too much experience with them but from the research it would seem that Quart would make a great candidate if you are migrating from Flask or want a very easy dev process. But when it comes to performance among all the modern ASGI frameworks it seems to be towards the bottom while Sanic tends to be near the top (well all this to be taken with a pinch-o-salt). The 3rd contender to throw in would be FastAPI - they have some very neat features and it seems very easy.... also near top performers.

FastAPI key features:

- Fast: Very high performance, on par with NodeJS and Go (thanks to Starlette and Pydantic). One of the fastest Python frameworks available.
- Fast to code: Increase the speed to develop features by about 200% to 300%. *
- Fewer bugs: Reduce about 40% of human (developer) induced errors. *
- Intuitive: Great editor support. Completion everywhere. Less time debugging.
- Easy: Designed to be easy to use and learn. Less time reading docs.
- Short: Minimize code duplication. Multiple features from each parameter declaration. Fewer bugs.
- Robust: Get production-ready code. With automatic interactive documentation.
- Standards-based: Based on (and fully compatible with) the open standards for APIs: OpenAPI (previously known as Swagger) and JSON Schema.
 
a very interesting nugget to [read here](https://devopedia.org/asynchronous-programming-in-python) which basically shows how crazy the uvloop + httptools performance can be.

**Comment by TheHolyRoger** [comment]
FastAPI is definitely a strong contender too.

@krisj my understanding of sanic is that ASGI is still new and unfinished, it's still migrating over from WSGI

**Comment by cardosofede** [comment]
@mifeng do you think that an script example of how to do this could be useful? Or you are looking for a module that updates the data of some api like the rate oracle? (I could be the two options)

**Comment by mifeng** [comment]
I think we should wait until we have a proper discussion about this in a forum thread when the new Discourse forum is up, since the CoinAlpha team also has discussed this idea internally, and other community members should also have a chance to chime in. We can also raise this topic in one of the early Weekly Developer Calls.

For script examples, I think we should stick with ones that use the existing order book data.

**Comment by MementoRC** [comment]
Are we closer to a framework for this? Could microservice with the new orchestration framework be a part of it?

**Comment by fengtality** [comment]
The new Candles Feed which supports ta-lib indicators handles this. What
additional features are needed?

**Comment by MementoRC** [comment]
My mistake, I misunderstood the scope with `signal` as in any type of messaging, thus the orchestration and microservice reference.
I agree with you. So we should close this issue satisfactorily
Thank you
